### ISSUE TITLE
Reverse proxy and IIS integrated security

### ISSUE THREAD

**Issue by molinch** [nghttp2/nghttp2#213]
When nghttpx is configured as a reverse proxy to an IIS backend using integrated security (ie: Kerberos) then the browser will keep asking for credentials.
It looks like something is not passed from browser to backend server.

Accessing the backend server directly will smoothly authenticate: not a single popup, expected behaviour.
By any chance do you know a workaround?

**Comment by tatsuhiro-t** [comment]
Possible cause of this issue is header field is too large.  Kerberos's header is known to be large, ~64KB or greater.
Currently, nghttpx limits 32KB for whole header field set for both frontend and backend.
This is hard coded value, and you can change this to large value, say 1MB:
```diff
diff --git a/src/shrpx_downstream.h b/src/shrpx_downstream.h
index 1f55cd4..1a24ca8 100644
--- a/src/shrpx_downstream.h
+++ b/src/shrpx_downstream.h
@@ -287,7 +287,7 @@ public:
   int change_priority(int32_t pri);
 
   // Maximum buffer size for header name/value pairs.
-  static const size_t MAX_HEADERS_SUM = 32768;
+  static const size_t MAX_HEADERS_SUM = 1024 * 1024;
 
   bool get_rst_stream_after_end_stream() const;
   void set_rst_stream_after_end_stream(bool f);
```

**Comment by molinch** [comment]
I did this change but still have the issue.

I also changed IIS settings to use basic authentication but still the popup keeps coming multiple time.
With Firefox and direct access: 1 popup.
With Firefox and reverse proxy: 10 or 15 popups.

**Comment by molinch** [comment]
The strange thing is that I don't have these issues when using trusterd reverse proxy.
But... trusterd is using nghttp2 under the hood.

**Comment by tatsuhiro-t** [comment]
We are in the same situation regarding NTLM with nginx:
http://forum.nginx.org/read.php?2,72871,220959#msg-220959

It is interesting to hear that trusterd works with NTLM.  Probably, it has longer keep-alive timeout in backend.  nghttpx by default very small keep-alive timeout and you can make it longer but nghttpx reuses backend connections among many frontend connections, which is not suitable with NTLM.

libnghttp2 is only HTTP/2 framing layer only, and application has to implement networking layer.

**Comment by oyvindhorneland** [comment]
As this is Kerberos and not NTLM it might not be the same issue.

@molinch Could you provide a trace from tcpdump/wireshark that shows traffic from client => nghttpx => backend (provided it's not encrypted)? If encrypted from client to nghttpx, a trace from Fiddler2 will do for that conenction. Let's see if the negotiate Authorization was sent correctly and that it survived intact all the way to the backend server. 

Do you have any messages from the backend server regarding why it fails to verify the presented credentials (possible to enable kerberos debug logging)?

**Comment by tatsuhiro-t** [comment]
oh, sorry I mixed up NTLM and kerberos...

**Comment by oyvindhorneland** [comment]
@tatsuhiro-t As the kerberos ticket may exceed the current max size, could you make the max size a configuration option?

I've seen 48000 used in production so far and a search on kerberos ticket suggests that 48000 might be a common max setting, although it could even be larger than that...
https://support.microsoft.com/en-us/kb/327825
http://blogs.technet.com/b/shanecothran/archive/2010/07/16/maxtokensize-and-kerberos-token-bloat.aspx
https://support.microsoft.com/en-us/kb/938118

**Comment by tatsuhiro-t** [comment]
The limit of sum of header field size was extended to 128KiB recently, which is large enough for kerberos ticket.
If you think that is too large (way larger than 48KiB), we can add option (to nghttpx) to let user set the value they think reasonable.

**Comment by oyvindhorneland** [comment]
Apache's LimitRequestFieldSize (default 8192):
> This directive gives the server administrator greater control over abnormal client request behavior, which may be useful for avoiding some forms of denial-of-service attacks.

Jetty org.eclipse.jetty.server.HttpConfiguration (default 8192):
> public void setRequestHeaderSize(int requestHeaderSize)
> Set the maximum size of a request header.
> Larger headers will allow for more and/or larger cookies plus larger form content encoded in a URL. However, larger headers consume more memory and can make a server more vulnerable to denial of service attacks.
Parameters:
requestHeaderSize - Max header size in bytes

It seems a configurable option is the recommended way to go to also be able to harden against DoS attacks.

**Comment by tatsuhiro-t** [comment]
Right. Do you think we also use 8KiB as default value? Or 48KiB+ to support kerberos by default?

**Comment by tatsuhiro-t** [comment]
Further reading apache configuration, it seems that LimitRequestFieldSize controls the number of bytes per single Http request header field.  Apache also has LimitRequestFields, which controls the maximum number of header fields to accept.  This directive defaults to 100.
If we are going to mimic apache configuration options, we need these 2 options.

**Comment by tatsuhiro-t** [comment]
I decided to go with 2 options: 1) specify maximum buffer size to receive whole request header field set, which is the sum of name and value in bytes, and 2) the number of header fields allowed in a header field set.

**Comment by oyvindhorneland** [comment]
Sounds great! A default value of 128KiB is probably fine for the complete header set.

**Comment by tatsuhiro-t** [comment]
I just went in conservative mode and use 64KiB as default buffer size, since it can cope with 48KiB kerberos header plus some cookies.
The fix was already committed to master.  The added options are:
```
  --header-field-buffer=<SIZE>
              Set maximum  buffer size for incoming  HTTP header field
              list.   This is  the sum  of  header name  and value  in
              bytes.
              Default: 64K
  --max-header-fields=<N>
              Set maximum number of incoming HTTP header fields, which
              appear in one request or response header field list.
              Default: 100
```

**Comment by oyvindhorneland** [comment]
OK and should definitely work for most cases.
